src/charm.py

```diff
@@ -1,29 +1,130 @@
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
 from ops.pebble import ChangeError
 
-import urllib.parse
-import requests
-import yaml
-import json
-import logging
-from typing import Dict, List, Optional, Any
+import utils
 
 logger = logging.getLogger(__name__)
 
 
+class K8sServicePatch:
+    namespace_file = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+
+    class PatchFailed(RuntimeError):
+        pass
+
+    @staticmethod
+    def namespace() -> str:
+        """Read the Kubernetes namespace we're deployed in from the mounted service token
+
+        Returns:
+            str: The current Kubernetes namespace
+        """
+        with open(K8sServicePatch.namespace_file, "r") as f:
+            return f.read().strip()
+
+    @staticmethod
+    def _k8s_auth():
+        """Authenticate with the Kubernetes API using an in-cluster service token
+
+        Raises:
+            K8sServicePatch.PatchFailed: if no permissions to read cluster role
+        """
+        # Authenticate against the Kubernetes API using a mounted ServiceAccount token
+        kubernetes.config.load_incluster_config()
+        # Test the service account we've got for sufficient perms
+        api = kubernetes.client.CoreV1Api(kubernetes.client.ApiClient())
+
+        try:
+            api.list_namespaced_service(namespace=K8sServicePatch.namespace())
+        except kubernetes.client.exceptions.ApiException as e:
+            if e.status == 403:
+                raise K8sServicePatch.PatchFailed(
+                    "No permission to read cluster role. " "Run `juju trust` on this application."
+                ) from e
+            else:
+                raise e
+
+    @staticmethod
+    def _k8s_service(
+        app: str, service_ports: List[Tuple[str, int, int]]
+    ) -> kubernetes.client.V1Service:
+        """Property accessor to return a valid Kubernetes Service representation for Alertmanager
+
+        Args:
+            app: app name
+            service_ports: a list of tuples (name, port, target_port) for every service port.
+
+        Returns:
+            kubernetes.client.V1Service: A Kubernetes Service with correctly annotated metadata and ports
+        """
+
+        ports = [
+            kubernetes.client.V1ServicePort(name=port[0], port=port[1], target_port=port[2])
+            for port in service_ports
+        ]
+
+        ns = K8sServicePatch.namespace()
+        return kubernetes.client.V1Service(
+            api_version="v1",
+            metadata=kubernetes.client.V1ObjectMeta(
+                namespace=ns,
+                name=app,
+                labels={"app.kubernetes.io/name": app},
+            ),
+            spec=kubernetes.client.V1ServiceSpec(
+                ports=ports,
+                selector={"app.kubernetes.io/name": app},
+            ),
+        )
+
+    @staticmethod
+    def patch(app: str, service_ports: List[Tuple[str, int, int]]):
```

Comment:
If this function is meant only to patch a port, wouldn't it be better to call it something like `patch_port()`.

Reply:
Perhaps `set_ports`?

Also as discussed yesterday, perhaps a class called `PortMapping` with properties for `from_port`, `to_port`, `name` with the `__repr__` and `__dict__` overriden here could result in tider type hints?

Reply:
Done